#summary Proposals for modeling sequence features in RDF/OWL

= Proposals for Modeling Sequence Features in RDF/OWL =

The motivation for these RDF/OWL models is the development of [ServicesForGMOD SADI services for GMOD].  

= Genomic Regions =

A genomic region is a contiguous chain of bases/amino acids within a biological sequence (e.g. a chromosome or a protein).  

A genomic region is described by a combination of:

  * a *start position*: the index of the first base / amino acid
  * an *end position*: the index of the last base / amino acid
  * a *strand*: the strand ('+', '-', or 'unknown')
  * a *reference feature*: determines the frame of reference for the  coordinates (start, end, and strand)  
  * a *genome assembly*: the genome that contains the genomic region

Each coordinate of the genomic range is optional:

  * *no start position* means that the genomic range starts at the first base / amino acid of the reference feature
  * *no end position* means that the genomic range ends at the last base / amino acid of the reference feature
  * *no strand* means that the strand is irrelevant (e.g. a region of a protein)

We adopt the same semantics for "strand" as GMOD. The strand part of the genomic region has three possible values:

  * +1: the feature is on the same DNA/RNA strand as the reference feature and has the same direction 
  * -1: the feature is on the opposite strand of the reference feature and has the opposite direction
  *  0: the strand is relevant, but unknown.

For example, if we have a short chromosome consisting of:

AGCTAGCTAGCT

then a region with coordinates (start,end,strand) = (2,6,-1) has the sequence: 

CTAGC

= Features =

A feature is a typed genomic region. Examples of features are chromosomes, genes, transcripts, exons, transcription factor binding sites, proteins, protein domains, and so on.  Everything is a feature.  

Every feature has:

  * a *genomic region*, as described above
  * an *identifier* that is unique to the sequence database (e.g. "chrIX" for the ninth chromosome in yeast)

== Example Feature in RDF/N3 ==

Suppose the example feature has ID 'BindingSite1', and the ID of the chromosome is 'ChrIX'. Then the feature could be encoded in RDF/N3 as:

{{{
@prefix feature: <http://sadiframework.org/ontologies/GMOD/feature.owl#>
@prefix region: <http://sadiframework.org/ontologies/GMOD/genomic_region.owl#>
@prefix assembly: <http://sadiframework.org/ontologies/GMOD/genome_assembly.owl#>

@prefix featureURI: <TO_BE_DECIDED>
@prefix regionURI: <TO_BE_DECIDED>
@prefix assemblyURI: <TO_BE_DECIDED>

@prefix sio: <http://semanticscience.org/resource/>
@prefix so: <http://purl.org/obo/owl/SO#>

@prefix xsd: <http://www.w3.org/2001/XMLSchema#>

featureURI:BindingSite1
     a feature:Feature; 
     a so:SO_0000235;                                        # 'TF_binding_site'
     sio:SIO_000008                                          # 'has attribute' 
        [ a sio:SIO_000115;                                  # 'identifier'
          sio:SIO_000300 'BindingSite1'^^xsd:string ];
     sio:SIO_000061                                          # 'is located in'
        [ a region:GenomicRegion;
          region:is-relative-to featureURI:Chr1;             # reference feature
          region:SIO_000061 assemblyURI:Assembly1.2.0;       # 'is located in'
          region:start-position '2'^^xsd:nonNegativeInteger;
          region:end-position '6'^^xsd:nonNegativeInteger;
          region:strand '-1'^^xsd:integer; ] .
       
feature:Chr1
     a feature:Feature; 
     a so:SO_0000340;                                        # 'chromosome'
     sio:SIO_000008                                          # 'has attribute' 
        [ a sio:SIO_000115;                                  # 'identifier'
          sio:SIO_000300 'Chr1'^^xsd:string ];
     sio:SIO_000061                                          # 'is located in'
        [ a region:GenomicRegion;
          region:is-relative-to featureURI:Chr1;             # reference feature
          region:SIO_000061 assemblyURI:Assembly1.2.0; ] .   # 'is located in'
            
assemblyURI:Assembly1.2.0
     a assembly:Assembly;
     sio:SIO_000066                                          # 'has source'
        [ a assembly:Organism;
          sio:SIO_000008                                     # 'has attribute' 
              [ a lsrn:taxon_Identifier;
                sio:SIO_000300 '4932'^^xsd:string; ]
     assembly:version '1.2.0'^^xsd:string .                 

}}}    