#summary Proposals for modeling sequence features in RDF/OWL

= Proposals for Modeling Sequence Features in RDF/OWL =

The motivation for these RDF/OWL models is the development of [ServicesForGMOD SADI services for GMOD].  

= Features =

A feature is an annotated region of a biological sequence (e.g. a chromosome or an amino acid sequence). Examples of features are chromosomes, genes, transcripts, exons, transcription factor binding sites, proteins, protein domains, and so on.  Everything is a feature.  

Every feature has:

  * a *genomic range*, which is described by any combination of a *start position*, an *end position*, and a *strand*. (Each part is optional; see below.)
  * a *reference feature*, which determines the reference frame for the above genomic range
  * an *identifier* that is unique to the sequence database (e.g. "chrIX" for the 9th chromosome in yeast)

Top-level features in the sequence database (usually a chromosome) point to themselves as a reference feature (example below). 

We adopt the same semantics for "strand" as GMOD. The strand part of the genomic range has a value of either +1 or -1.  In the first case (+1), the feature is on the same DNA/RNA strand as the reference feature and has the same direction. In the second case (-1), the feature is on the opposite strand of the reference feature and has the opposite direction. For example, if we have a short chromosome consisting of:

AGCTAGCTAGCT

then a feature with coordinates (start,end,strand) = (2,6,-1) has the sequence: 

CTAGC

Each coordinate of the genomic range is optional:

  * *no start position* means that the genomic range starts at the first base / amino acid of the reference feature
  * *no end position* means that the genomic range ends at the last base / amino acid of the reference feature
  * *no strand* means one of three things: i) the feature is on the same strand as the reference feature, ii) the strand is irrelevant (e.g. region of a domain in a protein), iii) the strand is relevant but unknown.

== Example Feature in RDF/N3 ==

Suppose the example feature has ID 'BindingSite1', and the ID of the chromosome is 'ChrIX'. Then the feature could be encoded in RDF/N3 as:

`@prefix feature: <http://sequencedb.org/feature/>
@prefix so: <http://

feature:BindingSite1 
     a so ;
     a 

`    